File: solutions/java/isbn-verifier/1/src/main/java/IsbnVerifier.java
import java.util.stream.IntStream;

class IsbnVerifier {

	boolean isValid(String stringToVerify) {
		if (stringToVerify == null || stringToVerify.isEmpty()) {
			return false;
		}

		String isbn = stringToVerify.replace("-", "");
		if (isbn.length() != 10) {
			return false;
		}

		try {
			int sum = this.getSum(isbn);
			return sum % 11 == 0;
		}
		catch (IllegalArgumentException e) {
			return false;
		}
	}

	private int getSum(String isbn) {
		return IntStream.range(0, 10)
				       .map(i -> this.weightedIsbnDigit(i, isbn))
				       .sum();
	}

	private int weightedIsbnDigit(int index, String isbn) {
		char character = isbn.charAt(index);
		if (index == 9 && character == 'X') {
			return 10 * (10 - index);
		}
		else if (Character.isDigit(character)) {
			return Character.getNumericValue(character) * (10 - index);
		}
		else {
			throw new IllegalArgumentException("Invalid character in ISBN");
		}
	}
}

File: solutions/python/isbn-verifier/1/isbn_verifier.py
ISBN_LENGTH = 10
CHECKSUM_MODULUS = 11
X_VALUE = 10
CHECK_CHARACTER = "X"
DASH = "-"

def _remove_dashes(isbn):
    return isbn.replace(DASH, "")

def _has_valid_characters(isbn):
    for character in isbn[:-1]:
        if not character.isdigit():
            return False

    return (
            isbn[-1].isdigit()
            or isbn[-1] == CHECK_CHARACTER
    )

def _get_check_value(check_character):
    if check_character == CHECK_CHARACTER:
        return X_VALUE

    return int(check_character)

def _calculate_checksum(isbn):
    checksum = 0
    for index, character in enumerate(isbn):
        weight = ISBN_LENGTH - index
        value = _get_check_value(character)
        checksum += value * weight

    return checksum

def is_valid(isbn):
    normalized_isbn = _remove_dashes(isbn)
    if len(normalized_isbn) != ISBN_LENGTH:
        return False

    if not _has_valid_characters(normalized_isbn):
        return False

    checksum = _calculate_checksum(normalized_isbn)
    return checksum % CHECKSUM_MODULUS == 0